test/test-runtime-onMessage.js

```diff
@@ -0,0 +1,165 @@
+"use strict";
+
+const {deepEqual, equal, ok} = require("chai").assert;
+const sinon = require("sinon");
+
+const {setupTestDOMWindow} = require("./setup");
+
+describe("browser-polyfill", () => {
+  describe("wrapped runtime.onMessage listener", () => {
+    it("do not wrap the listener if it is not a function", () => {
+      const fakeChrome = {
+        runtime: {
+          lastError: null,
+          onMessage: {
+            addListener: sinon.spy(),
+            hasListener: sinon.stub(),
+            removeListener: sinon.spy(),
+          },
+        },
+      };
+
+      return setupTestDOMWindow(fakeChrome).then(window => {
+        const fakeNonFunctionListener = {fake: "non function listener"};
+
+        window.browser.runtime.onMessage.addListener(fakeNonFunctionListener);
+
+        deepEqual(fakeChrome.runtime.onMessage.addListener.firstCall.args[0],
+                         fakeNonFunctionListener,
+                         "The non-function listener has not been wrapped");
+      });
+    });
+
+    it("keeps track of the listeners added", () => {
+      const messageListener = sinon.spy();
+      const fakeChromeListeners = new Set();
+      const fakeChrome = {
+        runtime: {
+          lastError: null,
+          onMessage: {
+            addListener: sinon.spy((listener, ...args) => {
+              fakeChromeListeners.add(listener);
+            }),
+            hasListener: sinon.spy(listener => fakeChromeListeners.has(listener)),
+            removeListener: sinon.spy(listener => {
+              fakeChromeListeners.delete(listener);
+            }),
+          },
+        },
+      };
+
+      return setupTestDOMWindow(fakeChrome).then(window => {
+        equal(window.browser.runtime.onMessage.hasListener(messageListener),
+                     false, "Got hasListener==false before the listener has been added");
+
+        window.browser.runtime.onMessage.addListener(messageListener);
+
+        equal(window.browser.runtime.onMessage.hasListener(messageListener),
+                     true, "Got hasListener==true once the listener has been added");
+
+        // Add the same listener again to test that it will be called with the
+        // same wrapped listener.
+        window.browser.runtime.onMessage.addListener(messageListener);
+
+        ok(fakeChrome.runtime.onMessage.addListener.calledTwice,
+                  "addListener has been called twice");
+        equal(fakeChrome.runtime.onMessage.addListener.secondCall.args[0],
+                     fakeChrome.runtime.onMessage.addListener.firstCall.args[0],
+                     "both the addListener calls received the same wrapped listener");
```

Comment:
We should also check that calling with a different function passes through a different wrapper, and so forth.